tests/database/odbc-w-enclave/odbc_proxy.c

```diff
@@ -0,0 +1,388 @@
+
+    result = SUCCESS;
+done:
+    return result;
+}
+
+static int query_table(SQLHSTMT stmt)
+{
+    int result = FAILURE;
+    SQLRETURN rc;
+    char data[DEFAULT_VCHAR_LEN];
+
+    rc = SQLExecDirect(stmt, "SELECT * FROM " TABLE_NAME, SQL_NTS);
+    OK_CHECK(checkRC(rc, "Query table", stmt, SQL_HANDLE_STMT));
+    rc = SQLBindCol(stmt, 1, SQL_C_CHAR, data, DEFAULT_VCHAR_LEN, 0);
+
+    if (SQL_SUCCESS == (rc = SQLFetch(stmt)))
+    {
+        printf("\nQuery from encrypted column successfully. Decrepted data: %s\n", data);
+    }
+
+    SQLFreeStmt(stmt, SQL_CLOSE);
+    OK_CHECK(checkRC(rc, "Free statement", stmt, SQL_HANDLE_STMT));
+
+    result = SUCCESS;
+done:
+    return result;
+}
+
+int odbc_proxy_connectDB(char* connstr)
+{
+    int result = FAILURE;
+    SQLRETURN rc;
+    char sqlbuf[SQL_BUFFER_LEN];
+    char* ckspData = "";
+    size_t ckspDataSize = sizeof(char*);
+
+    // Connect to Database Server
+    rc = SQLAllocHandle(SQL_HANDLE_ENV, NULL, &env);
+    OK_CHECK(checkRC(rc, "allocating environment handle", 0, 0));
+    rc = SQLSetEnvAttr(env, SQL_ATTR_ODBC_VERSION, (SQLPOINTER)SQL_OV_ODBC3, 0);
+    OK_CHECK(checkRC(rc, "setting ODBC version to 3.0", env, SQL_HANDLE_ENV));
+    rc = SQLAllocHandle(SQL_HANDLE_DBC, env, &dbc);
+    OK_CHECK(checkRC(rc, "allocating connection handle", env, SQL_HANDLE_ENV));
+    rc = SQLDriverConnect(
+        dbc, 0, connstr, strlen(connstr), NULL, 0, NULL, SQL_DRIVER_NOPROMPT);
+    OK_CHECK(checkRC(rc, "connecting to data source", dbc, SQL_HANDLE_DBC));
+
+    // load Custom Key Store Provider
+    rc = SQLSetConnectAttr(
+        dbc, SQL_COPT_SS_CEKEYSTOREPROVIDER, DLLPATH, SQL_NTS);
+    OK_CHECK(checkRC(rc, "Loading KSP into ODBC Driver", dbc, SQL_HANDLE_DBC));
+
+    // Write Custom Key Store Provider, this step is for SQLAEv2
+    // compatibility
+    if (ckspData != NULL && ckspDataSize > 0)
+    {
+        unsigned char ksd[sizeof(CEKEYSTOREDATA) + ckspDataSize];
+        CEKEYSTOREDATA* pKsd = (CEKEYSTOREDATA*)ksd;
+        pKsd->name = KSPNAME;
+        pKsd->dataSize = ckspDataSize;
+        memcpy(pKsd->data, ckspData, ckspDataSize);
+        rc = SQLSetConnectAttr(
+            dbc, SQL_COPT_SS_CEKEYSTOREDATA, (SQLPOINTER)pKsd, SQL_IS_POINTER);
+        OK_CHECK(checkRC(rc, "Configuring the KSP", dbc, SQL_HANDLE_DBC));
+    }
+
+    result = SUCCESS;
+done:
+    return result;
+}
+
+int odbc_proxy_execute()
+{
+    int result = FAILURE;
+    SQLHSTMT stmt;
+    SQLRETURN rc = SQLAllocHandle(SQL_HANDLE_STMT, dbc, &stmt);
+    OK_CHECK(checkRC(rc, "allocating statement handle", dbc, SQL_HANDLE_DBC));
+
+    // 1. before start drop all tables and CEKs, CMKs
+    OK_CHECK(clean_up(stmt));
+
+    // 2. Create CMK (Column Master Key)
+    OK_CHECK(create_CMK(stmt));
+
+    // 3. create CEK (Column Encryption Key)
+    OK_CHECK(create_ecek(stmt));
+
+    // 4. Create Data table
+    OK_CHECK(create_table(stmt));
```

Comment:
Would this fail if table already exits?

Reply:
Yes. But in step 1, we drop data table first, if it exists